File: apps/party-time-backend/src/main/java/com/partytime/configuration/security/AuthEntryPointJwt.java
package com.partytime.configuration.security;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.partytime.api.error.ApiError;
import jakarta.servlet.ServletException;
import jakarta.servlet.http.HttpServletRequest;
import jakarta.servlet.http.HttpServletResponse;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.http.HttpHeaders;
import org.springframework.http.MediaType;
import org.springframework.security.core.AuthenticationException;
import org.springframework.security.web.AuthenticationEntryPoint;
import org.springframework.stereotype.Component;

import java.io.IOException;
import java.io.PrintWriter;

@Slf4j
@Component
@RequiredArgsConstructor
public class AuthEntryPointJwt implements AuthenticationEntryPoint {

    private final ObjectMapper objectMapper;

    @Override
    public void commence(HttpServletRequest request, HttpServletResponse response,
                         AuthenticationException authException) throws IOException, ServletException {
        log.error("Unauthorized error: {}", authException.getMessage());
        response.setStatus(401);
        response.addHeader(HttpHeaders.CONTENT_TYPE, MediaType.APPLICATION_JSON_VALUE);
        ApiError error = ApiError.unauthorized();
        String json = objectMapper.writeValueAsString(error);
        PrintWriter writer = response.getWriter();
        writer.write(json);
        writer.flush();
    }


}

File: apps/party-time-backend-kotlin/src/main/kotlin/com/partytime/configuration/security/AuthEntryPointJwt.kt
package com.partytime.configuration.security

import com.fasterxml.jackson.databind.ObjectMapper
import com.partytime.api.error.ApiError
import io.github.oshai.kotlinlogging.KotlinLogging
import jakarta.servlet.http.HttpServletRequest
import jakarta.servlet.http.HttpServletResponse
import org.springframework.http.HttpHeaders
import org.springframework.http.HttpStatus
import org.springframework.http.MediaType
import org.springframework.security.core.AuthenticationException
import org.springframework.security.web.AuthenticationEntryPoint
import org.springframework.stereotype.Component

private val authLogger = KotlinLogging.logger {}
@Component
class AuthEntryPointJwt (
    private val objectMapper: ObjectMapper
): AuthenticationEntryPoint {


    override fun commence(
        request: HttpServletRequest,
        response: HttpServletResponse,
        authException: AuthenticationException
    ) {
        authLogger.error { "Unauthorized error: ${authException.message}" }
        response.apply {
            status = HttpStatus.UNAUTHORIZED.value()
            addHeader(HttpHeaders.CONTENT_TYPE, MediaType.APPLICATION_JSON_VALUE)

            val error = ApiError.unauthorized()
            val json = objectMapper.writeValueAsString(error)
            writer.write(json)
            writer.flush()
        }
    }
}
